### DISCUSSION TITLE
Minor issue in arguments for `AttributeImpl`

### DISCUSSION THREAD

**Discussion by indiVar0508** [sqlalchemy/sqlalchemy#10474] [Usage Questions] [answered]
Hi 
I am not sure if its a bug or if i am using a library that has issue so wanted to confirm here once,

In `orm.attribute.py` the method [`register_attribute_impl` ](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/attributes.py#L2620), arguments are passed as `class_, key, dispatch, ...`

```py3
    if impl_class:
        # TODO: this appears to be the WriteOnlyAttributeImpl /
        # DynamicAttributeImpl constructor which is hardcoded
        impl = cast("Type[WriteOnlyAttributeImpl]", impl_class)(
            class_, key, dispatch, **kw
        )
```

but in [`AttributeImpl`](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/attributes.py#L833) the constructor methods third argument is a `callable_` which is not being sent from `register_attribute_impl`
so it gives me error 

`TypeError: __init__() missing 1 required positional argument: 'dispatch'`

```py
    def __init__(
        self,
        class_: _ExternalEntityType[_O],
        key: str,
        callable_: Optional[_LoaderCallable],  # < - third argument.
        dispatch: _Dispatch[QueryableAttribute[Any]],
        trackparent: bool = False,
        ...
        )
```

Not sure, but if downgrade my sqlalchemy to 2.0.21 it works but with sqlalchemy 2.0.22 it fails.

**Comment by CaselIT** [comment]
Hi,

With what you have posted it's impossible to provide any help.

How are you getting that error? That code path is used for write only, that doesn't have callable as argument https://github.com/sqlalchemy/sqlalchemy/blob/d2ce313d3daeb2d8efd3adbd1d7e8d85796ee759/lib/sqlalchemy/orm/writeonly.py#L164.

Please provide an example of what you are trying to do, thanks.

**Reply by indiVar0508** [reply]
I see got it i just got confused that when i am running with sqla==2.0.21 it was working but with 2.0.22 it failed so i thought something changes and was tracing it.
Noticed now someone has raised issue to package also it might be the package needs to change.
https://github.com/kvesteri/sqlalchemy-utils/issues/719

closing this question, thanks.

**Reply by CaselIT** [reply]
It seems that sqlalchemy_utils is using an internal api in an unsupported way.

**Reply by CaselIT** [reply]
@zzzeek forgot to confirm: is `register_attribute_impl` private so it should not be reliead upon by 3rd party libs?

**Reply by zzzeek** [reply]
it's not marked private at the moment but certainly I treat this as private so I guess we have to underscore all these things to keep sqlalchemy-utils away.

**Reply by CaselIT** [reply]
ok. the path where `impl_class` is passed is only meant for dynamic / write-only.

would it be viable to create a public api for external libs?

**Reply by zzzeek** [reply]
to ...make their own kinds of mapped attributes ?   I haven't looked at sqlalchemy_utils.  is that what they're doing ?

**Reply by CaselIT** [reply]
apparently they define a so called [`generic_relationship`](https://github.com/kvesteri/sqlalchemy-utils/blob/master/sqlalchemy_utils/generic.py). I don't know what it does

**Reply by zzzeek** [reply]
it's probably what we have here: https://docs.sqlalchemy.org/en/20/_modules/examples/generic_associations/generic_fk.html 

but I really would have preferred they didnt make their own attribute impl

**Reply by indiVar0508** [reply]
Hi @CaselIT ,

`It seems that sqlalchemy_utils is using an internal api in an unsupported way.`
I was trying to look into the issue by going through traceback had 2 minor queries, 
1. here when constructor is being intialised for [writableAttrImpl](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/writeonly.py#L173) the callable_ attribute is being passed as None, but when [registring attr impl ](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/attributes.py#L2620)it is not being done? 
2. i checked the flow so they are calling attributes.register_attribute method is that also suppose to be private?

**Reply by CaselIT** [reply]
The api is used internally by sqlalchemy and there was no expectation of it being stable. The code that it calls impl has changed following the addition of types to some of the modules: 0a12d22d6d63bae1c695f0afc90f9cb6d825b82e

**Comment by marksteward** [comment, marked as answer]
Fixed by https://github.com/kvesteri/sqlalchemy-utils/pull/725
